client/fetch_state.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/encoding"
+	"github.com/m3db/m3db/serialize"
+	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3db/x/xpool"
+	"github.com/m3db/m3x/ident"
+)
+
+const (
+	maxUint = ^uint(0)
+	maxInt  = int(maxUint >> 1)
+)
+
+var (
+	errFetchStateStillProcessing = errors.New("[invariant violated] fetch state is still processing, unable to create response")
+)
+
+type fetchState struct {
+	sync.Cond
+	sync.Mutex
+	refCounter
+
+	op                   *fetchTaggedOp
+	tagResultAccumulator fetchResultsAccumulator
+	err                  error
+	done                 bool
+
+	pool fetchStatePool
+}
+
+func (f *fetchState) reset() {
+	f.destructorFn = f.close // Set refCounter completion as close
+	f.L = f                  // Set the embedded condition locker to the embedded mutex
+}
+
+func (f *fetchState) close() {
+	if f.op != nil {
+		f.op.decRef()
+		f.op = nil
+	}
+	f.err = nil
+	f.done = false
+	f.tagResultAccumulator.Clear()
+
+	if f.pool == nil {
+		return
+	}
+	f.pool.Put(f)
+}
+
+func (f *fetchState) Reset(
+	startTime time.Time,
+	endTime time.Time,
+	op *fetchTaggedOp,
+	topoMap topology.Map,
+	majority int,
+	consistencyLevel topology.ReadConsistencyLevel,
+) {
+	op.incRef() // take a reference to the provided op
+	f.op = op
+	f.tagResultAccumulator.Reset(startTime, endTime, topoMap, majority, consistencyLevel)
+}
+
+func (f *fetchState) completionFn(
+	result interface{},
+	resultErr error,
+) {
+	f.Lock()
+	defer func() {
+		f.Unlock()
+		f.decRef() // release ref held onto by the hostQueue (via op.completionFn)
```

Comment:
Seems like this function will be called (potentially) multiple times, are multiple decRef calls ok?

Reply:
Yep. It will be called by each of the different hostQueues, hence the corresponding `incRef()` when we enqueue an op (refer `fetchTaggedAttemptWithRLock` in session.go )